okhttp-urlconnection/src/test/java/okhttp3/JavaNetAuthenticatorTest.java

```diff
@@ -0,0 +1,177 @@
+package okhttp3;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import javax.net.SocketFactory;
+import java.net.PasswordAuthentication;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.UnknownHostException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static java.net.HttpURLConnection.HTTP_PROXY_AUTH;
+import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+import static java.net.InetSocketAddress.createUnresolved;
+import static java.net.Proxy.Type.HTTP;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static java.util.UUID.randomUUID;
+import static okhttp3.Credentials.basic;
+import static okhttp3.Protocol.HTTP_1_1;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+@RunWith(Parameterized.class)
+public class JavaNetAuthenticatorTest {
+  private static JavaNetAuthenticator javaNetAuthenticator = new JavaNetAuthenticator();
+
+  @Parameter public List<String> challenges;
+  @Parameter(1) public String realmToAnswer;
+  @Parameter(2) public boolean proxyAuthenticate;
+
+  private int responseCode;
+  private String authenticateHeaderName;
+  private String authorizationHeaderName;
+  private String username = randomUUID().toString();
+  private String password = randomUUID().toString();
+  private String credentials = basic(username, password);
+  private Proxy proxy = new Proxy(HTTP, createUnresolved("localhost", 80));
+  private Route route = new Route(
+      new Address("localhost",
+          80,
+          Dns.SYSTEM,
+          SocketFactory.getDefault(),
+          null,
+          null,
+          null,
+          Authenticator.NONE,
+          null,
+          Collections.<Protocol>emptyList(),
+          Collections.<ConnectionSpec>emptyList(),
+          ProxySelector.getDefault()),
+      proxy,
+      createUnresolved("localhost", 80));
+
+  @Parameters(name = "challenges: {0}, realmToAnswer: \"{1}\", proxyAuthenticate: {2}")
+  public static Collection<Object[]> parameters() {
```

Comment:
Hardly, as I also use the parameters in the `@Before` method.
And it makes the test class a multitude longer.
I can call the setup method manually too of course.

But may I first ask why you think it is "super awkward", especially as I don't think it will be any different with a `check()` method with parameters.

Whether you have a `check()` method with parameters or a parametrized test like currently, if you want to set a breakpoint in the check method for one case, you set a a breakpoint and add a condition, checking for the parameters / fields, it is the same.

Alternatively you can set the breakpoint and only run the single test from the IDE, but this is also easily possible with parameterized tests. You can just right click the instance you want to re-run in the list of executed tests and run only that version.

Reply:
I want to put a breakpoint on the 31st sample dats. How do I do that? The check() method can incorporate setup!

Reply:
It is the same whether with parameterized test or with helper method that all tests delegate to.

Either
- set a breakpoint on the line where you want it
- add the condition `singletonList("Digest, Basic ,,realm=\"foo\"").equals(challenges) && "foo".equals(realmToAnswer) && !proxyAuthenticate`
- run all tests

or
- run all tests
- set the breakpoint without condition
- right-click the instance you want to debug and only run that instance

or
- comment out all parameter sets except the one you want to debug
- set the breakpoint without condition
- run all tests

The only advantage when having a helper method instead of a parameterized test is, if you want to debug a test from front to end, meaning you set the breakpoint in the actual test class and then step through until the line you are interested in, or you set a non-suspending breakpoint in the actual test method and then have a breakpoint with condition that this other breakpoint was hit. But I don't think that the latter is so much easier / less awkward than to have a conditional breakpoint that tests for the parameters.

Reply:
So, how to resolve it?
Should I keep it like @yschimke likes it or change it to what @swankjesse requested?